src/hotspot/share/memory/arena.cpp

```diff
@@ -44,22 +44,23 @@ STATIC_ASSERT(is_aligned((int)Chunk::non_pool_size, ARENA_AMALLOC_ALIGNMENT));
 //--------------------------------------------------------------------------------------
 // ChunkPool implementation
 
-// MT-safe pool of chunks to reduce malloc/free thrashing
+// MT-safe pool of same-sized chunks to reduce malloc/free thrashing
 // NB: not using Mutex because pools are used before Threads are initialized
-class ChunkPool: public CHeapObj<mtInternal> {
+class ChunkPool {
   Chunk*       _first;        // first cached Chunk; its first word points to next chunk
   size_t       _num_chunks;   // number of unused chunks in pool
-  size_t       _num_used;     // number of chunks currently checked out
-  const size_t _size;         // size of each chunk (must be uniform)
+  const size_t _size;         // (inner payload) size of the chunks this pool serves
 
   // Our four static pools
-  static ChunkPool* _large_pool;
-  static ChunkPool* _medium_pool;
-  static ChunkPool* _small_pool;
-  static ChunkPool* _tiny_pool;
+  static const int _num_pools = 4;
+  static ChunkPool _pools[_num_pools];
 
-  // return first element or null
-  void* get_first() {
+ public:
+  ChunkPool(size_t size) : _first(NULL), _num_chunks(0), _size(size) {}
+
+  // Remove a chunk from the pool and return it; NULL if pool is empty.
+  Chunk* remove_chunk() {
```

Comment:
I renamed ChunkPool::allocate() because it does not allocate anymore, it just stores existing chunks.

Is it just the naming, or do you also prefer that the old ChunkPool::allocate() actually allocated the Chunks (it called os::malloc)? I found that unnecessary for a pure cache and it duplicated the allocation code in Chunk::operator new().

Reply:
It's the names that I'm not a fan of.  The new implementation is good.  And it _is_ allocating, from a free-list.  I think it's pretty common for a free-list allocator design to have "failure" leave it to the client to decide what to do, rather than having the fallback built into the free-list allocator.

Reply:
Okay, I rename it back. Thanks for confirming!